lightning/src/util/persist.rs

```diff
@@ -250,3 +261,942 @@ where
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"error writing channel monitor {}",
+					e
+				);
+				chain::ChannelMonitorUpdateStatus::PermanentFailure
+			},
+		}
+	}
+	
+
+	/// Persists a channel update, writing only the update to the provided
+	/// [`KVStore`] if possible.
+	///
+	/// In some cases, this will forward to [`MonitorUpdatingPersister::persist_new_channel`]:
+	///   
+	///   - The number of pending updates exceeds
+	///   [`self.maximum_pending_updates`]
+	///   - LDK will command re-persisting the entire monitor through this
+	/// function, specifically when `update` is `None`.
+	///   - If the update is at [`CLOSED_CHANNEL_UPDATE_ID`]
+	///   - If we don't know the last time a full monitor was persisted (edge
+	///   case)
+	fn update_persisted_channel(
+		&self,
+		funding_txo: OutPoint,
+		update: Option<&ChannelMonitorUpdate>,
+		monitor: &ChannelMonitor<ChannelSigner>,
+		update_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		if let Some(update) = update {
+			if update.update_id != CLOSED_CHANNEL_UPDATE_ID {
+				if let Some(last_monitor_update_id) = self
+					.last_monitor_update_id
+					.borrow()
+					.get(&MonitorName::from(funding_txo))
+				{
+					if update.update_id - last_monitor_update_id < self.maximum_pending_updates {
+						// This is an update to the monitor, which we persist to
+						// be applied on restart. 
+						//
+						// IMPORTANT: update_id: [`MonitorUpdateId`] is not to
+						// be confused with [`ChannelMonitorUpdate`]'s
+						// update_id.
+						let ns = match MonitorName::try_from(funding_txo) {
+							Ok(monitor_name) => self.monitor_update_namespace(&monitor_name),
+							Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+						};
+						let update_name = match UpdateName::try_from(update.update_id) {
+							Ok(update_name) => update_name,
+							Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+						};
+						let key = update_name.as_str();
+						return match self.kv_store.write(
+							CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+							&ns,
+							key,
+							&update.encode(),
+						) {
+							Ok(()) => chain::ChannelMonitorUpdateStatus::Completed,
+							Err(e) => {
+								log_error!(
+									self.logger,
+									"error writing channel monitor update {}",
+									e
+								);
+								chain::ChannelMonitorUpdateStatus::PermanentFailure
+							},
+						};
+					}
+				}
+			}
+		}
+		// We need to persist the entire new monitor and discard the old updates.
+		self.persist_new_channel(funding_txo, monitor, update_id)
+	}
+}
+
+/// A struct representing a name for a monitor.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct MonitorName(String);
+
+impl MonitorName {
+	/// Constructs a [`MonitorName`], after verifying that an [`OutPoint`] can
+	/// be formed from the given `name`.
+	pub fn new(name: String) -> Result<Self, io::Error> {
+		MonitorName::do_try_into_outpoint(&name)?;
```

Comment:
Well, this isn't just an assertion for debug/test builds, this is a real runtime validation. In an earlier comment, @TheBlueMatt had asked that I validate in the constructor(s) for monitor and update names. So the idea here is that you'd get one of the `InvalidData` errors here if you tried to construct a `MonitorName` that won't work. 

Technically, one needn't validate here since failure is inevitable as you'll get a malformed outpoint so nothing else will work, I guess doing this is just makes a stack trace for the proximal cause.